File: server/core/formatters.js
export function htmlToMarkdown(html) {
  let md = html;
  md = md.replace(/<h1[^>]*>(.*?)<\/h1>/gi, "# $1\n\n");
  md = md.replace(/<h2[^>]*>(.*?)<\/h2>/gi, "## $1\n\n");
  md = md.replace(/<h3[^>]*>(.*?)<\/h3>/gi, "### $1\n\n");
  md = md.replace(/<h4[^>]*>(.*?)<\/h4>/gi, "#### $1\n\n");
  md = md.replace(/<h5[^>]*>(.*?)<\/h5>/gi, "##### $1\n\n");
  md = md.replace(/<h6[^>]*>(.*?)<\/h6>/gi, "###### $1\n\n");
  md = md.replace(/<p[^>]*>(.*?)<\/p>/gi, "$1\n\n");
  md = md.replace(/<br\s*\/?>/gi, "\n");
  md = md.replace(/<strong[^>]*>(.*?)<\/strong>/gi, "**$1**");
  md = md.replace(/<b[^>]*>(.*?)<\/b>/gi, "**$1**");
  md = md.replace(/<em[^>]*>(.*?)<\/em>/gi, "*$1*");
  md = md.replace(/<i[^>]*>(.*?)<\/i>/gi, "*$1*");
  md = md.replace(/<code[^>]*>(.*?)<\/code>/gi, "`$1`");
  md = md.replace(/<pre[^>]*>(.*?)<\/pre>/gis, "```\n$1\n```\n");
  md = md.replace(/<a[^>]*href="([^"]*)"[^>]*>(.*?)<\/a>/gi, "[$2]($1)");
  md = md.replace(/<li[^>]*>(.*?)<\/li>/gi, "- $1\n");
  md = md.replace(/<blockquote[^>]*>(.*?)<\/blockquote>/gis, "> $1\n\n");
  md = md.replace(/<img[^>]*alt="([^"]*)"[^>]*src="([^"]*)"[^>]*\/?>/gi, "![$1]($2)");
  md = md.replace(/<hr\s*\/?>/gi, "---\n\n");
  md = md.replace(/<[^>]+>/g, "");
  md = md.replace(/&amp;/g, "&").replace(/&lt;/g, "<").replace(/&gt;/g, ">");
  md = md.replace(/&quot;/g, '"').replace(/&#39;/g, "'").replace(/&nbsp;/g, " ");
  md = md.replace(/\n{3,}/g, "\n\n").trim();
  return md;
}

export function formatOutput(data, format) {
  switch (format) {
    case "text":
      return data.text || "(no text content)";
    case "html":
      return data.html;
    case "markdown":
      return htmlToMarkdown(data.html);
    case "links":
      return data.links
        .map((l) => `${l.href}  ${l.text ? "| " + l.text : ""}`)
        .join("\n");
    case "json":
      return JSON.stringify(
        { title: data.title, url: data.url, text: data.text, links: data.links, meta: data.meta },
        null,
        2,
      );
    default:
      return data.text;
  }
}

export function formatGoogleResults(results, format) {
  switch (format) {
    case "json":
      return JSON.stringify(results, null, 2);
    case "links":
      return results.map((r) => r.url).join("\n");
    case "text":
      return results
        .map((r, i) => `${i + 1}. ${r.title}\n   ${r.url}\n   ${r.snippet}`)
        .join("\n\n");
    case "markdown":
      return results
        .map((r, i) => `${i + 1}. [${r.title}](${r.url})\n   > ${r.snippet}`)
        .join("\n\n");
    default:
      return JSON.stringify(results, null, 2);
  }
}

File: rust/src/formatters.rs
use regex::Regex;

use crate::models::{PageContent, SearchResultItem};

pub fn html_to_markdown(html: &str) -> String {
    let mut md = html.to_string();
    let replacements = [
        (r"(?i)<h1[^>]*>(.*?)</h1>", "# $1\n\n"),
        (r"(?i)<h2[^>]*>(.*?)</h2>", "## $1\n\n"),
        (r"(?i)<h3[^>]*>(.*?)</h3>", "### $1\n\n"),
        (r"(?i)<h4[^>]*>(.*?)</h4>", "#### $1\n\n"),
        (r"(?i)<h5[^>]*>(.*?)</h5>", "##### $1\n\n"),
        (r"(?i)<h6[^>]*>(.*?)</h6>", "###### $1\n\n"),
        (r"(?i)<p[^>]*>(.*?)</p>", "$1\n\n"),
        (r"(?i)<br\s*/?>", "\n"),
        (r"(?i)<strong[^>]*>(.*?)</strong>", "**$1**"),
        (r"(?i)<b[^>]*>(.*?)</b>", "**$1**"),
        (r"(?i)<em[^>]*>(.*?)</em>", "*$1*"),
        (r"(?i)<i[^>]*>(.*?)</i>", "*$1*"),
        (r"(?i)<code[^>]*>(.*?)</code>", "`$1`"),
        (r"(?is)<pre[^>]*>(.*?)</pre>", "```\n$1\n```\n"),
        (r#"(?i)<a[^>]*href="([^"]*)"[^>]*>(.*?)</a>"#, "[$2]($1)"),
        (r"(?i)<li[^>]*>(.*?)</li>", "- $1\n"),
        (r"(?is)<blockquote[^>]*>(.*?)</blockquote>", "> $1\n\n"),
        (
            r#"(?i)<img[^>]*alt="([^"]*)"[^>]*src="([^"]*)"[^>]*\/?>"#,
            "![$1]($2)",
        ),
        (r"(?i)<hr\s*/?>", "---\n\n"),
        (r"(?is)<[^>]+>", ""),
    ];

    for (pattern, replacement) in replacements {
        let regex = Regex::new(pattern).expect("valid formatter regex");
        md = regex.replace_all(&md, replacement).to_string();
    }

    md = md
        .replace("&amp;", "&")
        .replace("&lt;", "<")
        .replace("&gt;", ">")
        .replace("&quot;", "\"")
        .replace("&#39;", "'")
        .replace("&nbsp;", " ");

    Regex::new(r"\n{3,}")
        .expect("valid formatter regex")
        .replace_all(md.trim(), "\n\n")
        .to_string()
}

pub fn format_page_output(data: &PageContent, format: &str) -> String {
    match format {
        "text" => {
            if data.text.trim().is_empty() {
                "(no text content)".to_string()
            } else {
                data.text.clone()
            }
        }
        "html" => data.html.clone(),
        "markdown" => html_to_markdown(&data.html),
        "links" => data
            .links
            .iter()
            .map(|link| {
                if link.text.trim().is_empty() {
                    link.href.clone()
                } else {
                    format!("{}  | {}", link.href, link.text)
                }
            })
            .collect::<Vec<_>>()
            .join("\n"),
        "json" => serde_json::to_string_pretty(&serde_json::json!({
            "title": data.title,
            "url": data.url,
            "text": data.text,
            "links": data.links,
            "meta": data.meta,
        }))
        .expect("page content should serialize"),
        _ => data.text.clone(),
    }
}

pub fn format_search_results(results: &[SearchResultItem], format: &str) -> String {
    match format {
        "json" => serde_json::to_string_pretty(results).expect("results should serialize"),
        "links" => results
            .iter()
            .map(|result| result.url.clone())
            .collect::<Vec<_>>()
            .join("\n"),
        "text" => results
            .iter()
            .enumerate()
            .map(|(index, result)| {
                format!(
                    "{}. {}\n   {}\n   {}",
                    index + 1,
                    result.title,
                    result.url,
                    result.snippet
                )
            })
            .collect::<Vec<_>>()
            .join("\n\n"),
        "markdown" => results
            .iter()
            .enumerate()
            .map(|(index, result)| {
                format!(
                    "{}. [{}]({})\n   > {}",
                    index + 1,
                    result.title,
                    result.url,
                    result.snippet
                )
            })
            .collect::<Vec<_>>()
            .join("\n\n"),
        _ => serde_json::to_string_pretty(results).expect("results should serialize"),
    }
}
